lib/queue.js

```diff
@@ -946,139 +939,127 @@ Queue.prototype.getJob = function(jobId){
 };
 
 Queue.prototype.getFailedCount = function() {
-  return this.client.zcard(this.toKey('failed'));
+  return this.getJobCountByTypes('failed');
 };
 
 Queue.prototype.getDelayedCount = function() {
-  return this.client.zcard(this.toKey('delayed'));
+  return this.getJobCountByTypes('delayed');
 };
 
 Queue.prototype.getActiveCount = function() {
-  return this.client.llen(this.toKey('active'));
+  return this.getJobCountByTypes('active');
 };
 
 Queue.prototype.getWaitingCount = function() {
-  return this.client.llen(this.toKey('wait'));
+  return this.getJobCountByTypes('wait');
 };
 
 Queue.prototype.getPausedCount = function() {
-  return this.client.llen(this.toKey('paused'));
+  return this.getJobCountByTypes('paused');
 };
 
 Queue.prototype.getWaiting = function(start, end){
-  return Promise.join(
-    this.getJobs('wait', 'LIST', true, start, end),
-    this.getJobs('paused', 'LIST', true, start, end)).spread(function(waiting, paused){
-      return _.concat(waiting, paused);
-    });
+  return this.getJobs([ 'wait', 'paused' ], start, end, true);
 };
 
 Queue.prototype.getActive = function(start, end){
-  return this.getJobs('active', 'LIST', true, start, end);
+  return this.getJobs('active', start, end, true);
 };
 
 Queue.prototype.getDelayed = function(start, end){
-  return this.getJobs('delayed', 'ZSET', true, start, end);
+  return this.getJobs('delayed', start, end, true);
 };
 
 Queue.prototype.getCompleted = function(start, end){
-  return this.getJobs('completed', 'ZSET', false, start, end);
+  return this.getJobs('completed', start, end, false);
 };
 
 Queue.prototype.getFailed = function(start, end){
-  return this.getJobs('failed', 'ZSET', false, start, end);
+  return this.getJobs('failed', start, end, false);
 };
 
-Queue.prototype.getJobs = function(queueType, type, asc, start, end){
+Queue.prototype.getJobs = function(types, start, end, asc){
   var _this = this;
-  var key = this.toKey(queueType);
-  var jobs;
 
   start = _.isUndefined(start) ? 0 : start;
   end = _.isUndefined(end) ? -1 : end;
 
-  switch(type){
-    case 'LIST':
-      if(asc){
-        jobs = this.client.lrange(key, -(end + 1), -(start + 1)).then(function(result){
-          return result.reverse();
-        });
-      }else{
-        jobs = this.client.lrange(key, start, end);
-      }
-
-      break;
-    case 'ZSET':
-      jobs = asc ? this.client.zrange(key, start, end) : this.client.zrevrange(key, start, end);
-      break;
-  }
+  var resultByType = _this._commandByType(parseTypeArg(types), false, function(key, command){
+    switch(command){
+      case 'lrange':
+        if(asc){
+          return _this.client.lrange(key, -(end + 1), -(start + 1)).then(function(result){
+            return result.reverse();
+          });
+        }else{
+          return _this.client.lrange(key, start, end);
+        }
+      case 'zrange':
+        return asc ? _this.client.zrange(key, start, end) : _this.client.zrevrange(key, start, end);
+    }
```

Comment:
@manast I was initially hoping we'd be able to get rid of all this logic, or at least implement it as a transaction like we do with counts.

After working on this PR it became clear that a lot of what I was doing might be better encapsulated by a Lua script. If you think there's a better solution here let me know.